hilti/runtime/src/fiber.cc

```diff
@@ -1,95 +1,206 @@
+    // return the full stack size here which below leads to us copying way to
+    // much data and not saving any memory.
+    return ::fiber_stack_size(fbr);
+    // return ::fiber_stack_free_size(fbr);
+    // return ::fiber_stack_size(fbr);
+    // return ::fiber_stack_size(fbr) - ::fiber_stack_free_size(fbr);
+}
+
+struct PrepareSwitchArgs {
+    ::Fiber* fiber_switch_trampoline = nullptr;
+    ::Fiber* from = nullptr;
+    ::Fiber* to = nullptr;
+};
+
+void fiber_switch_finalize(void* argsp) {
+    auto* args = reinterpret_cast<PrepareSwitchArgs*>(argsp);
+    auto* fiber_switch_trampoline = args->fiber_switch_trampoline;
+    auto* from = args->from;
+    auto* to = args->to;
+
+    // Swap out `from` stack.
+    if ( from->stack ) {
+        auto stored_stack = globalState()->fiber_stacks.find(from);
+        if ( stored_stack == globalState()->fiber_stacks.end() ) {
+            auto [it, inserted] =
+                globalState()->fiber_stacks.insert_or_assign(from, Stack(detail::fiber_stack_size(from)));
+
+            HILTI_RT_DEBUG("fibers", fmt("added stack for %p, size=%s (%d)", from, detail::fiber_stack_size(from),
+                                         detail::fiber_stack_size(from) /
+                                             static_cast<double>(globalState()->fiber_shared_stack._size)));
+
+            assert(inserted);
+            stored_stack = it;
+        }
+        assert(stored_stack != globalState()->fiber_stacks.end());
+
+        auto& stack = stored_stack->second;
+        assert(stack._stack);
+        assert(stack._size > 0);
+
+        auto size = detail::fiber_stack_size(from);
+
+        if ( stack._size < size ) {
+            HILTI_RT_DEBUG("fibers", fmt("resizing fiber stack for %p from %s to %s", from, stack._size, size));
+            stack.resize(size);
+        }
+        assert(stack._size > 0);
+
+        HILTI_RT_DEBUG("fibers", fmt("storing stack for %p at %p", from, from->stack));
+        std::memcpy(stack._stack.get(), static_cast<char*>(from->stack) - size + from->stack_size, size);
+    }
+    else
+        HILTI_RT_DEBUG("fibers", fmt("not copying stack for from=%p since it is the main_co", from));
+
+    // If the fiber ran before swap in its stack.
+    if ( to->stack ) {
+        HILTI_RT_DEBUG("fibers", fmt("restoring stack for %p", to));
+        if ( auto stored_stack = globalState()->fiber_stacks.find(to);
+             stored_stack != globalState()->fiber_stacks.end() )
+            std::memcpy(static_cast<char*>(to->stack) - detail::fiber_stack_size(to) + to->stack_size,
+                        stored_stack->second._stack.get(), stored_stack->second._size);
+        else
+            HILTI_RT_DEBUG("fibers", fmt("stack for to=%p not found", to));
+    }
+    else
+        HILTI_RT_DEBUG("fibers", fmt("not copying stack for to=%p since it is the main_co", to));
+
+
+    return ::fiber_switch(fiber_switch_trampoline, to);
```

Comment:
This needs additional handling for ASAN. We probably need to move `Fiber`'s `asan` parameters into a subclass encapsulating both the `::Fiber` and the `asan` parameters.

Similarly for `fiber_switch` before it dispatches to this function.

Reply:
Currently the following tests fail:

```
spicy.types.sink.reassembler.gap
spicy.types.sink.reassembler.basic
spicy.types.sink.reassembler.skip
spicy.types.sink.reassembler.init-seq
spicy.types.sink.reassembler.overlap
spicy.types.unit.filter-chained
spicy.types.unit.params-forward
```

```
RESULT FILE .tmp/spicy.lib.protocols.http.gc-release/.stderr:
/bin/sh: line 1: 37803 Broken pipe: 13         awk -f generate.awk
     37804 Bus error: 10           | HILTI_DEBUG=spicy-driver spicy-driver http.hlto -i 83 -p HTTP::Replies > output 2> debug.log

RESULT FILE .tmp/spicy.types.sink.reassembler.basic/.stderr:
/bin/sh: line 1: 38973 Segmentation fault: 11  spicy-driver -p Mini::Main /Users/bbannier/src/spicy/tests/.tmp/spicy.types.sink.reassembler.basic/basic.spicy > output < /dev/null

RESULT FILE .tmp/spicy.types.sink.reassembler.gap/.stderr:
[libhilti] Internal error: code is executing that should not be reachable

--- Aborting in libhilti
/bin/sh: line 1: 58349 Abort trap: 6           (core dumped) spicy-driver -d -p Mini::Main /Users/bbannier/src/spicy/tests/.tmp/spicy.types.sink.reassembler.gap/gap.spicy > output < /dev/null

RESULT FILE .tmp/spicy.types.sink.reassembler.init-seq/.stderr:
/bin/sh: line 1: 39084 Segmentation fault: 11  spicy-driver -p Mini::Main /Users/bbannier/src/spicy/tests/.tmp/spicy.types.sink.reassembler.init-seq/init-seq.spicy > output < /dev/null

RESULT FILE .tmp/spicy.types.sink.reassembler.overlap/.stderr:
[libhilti] Internal error: code is executing that should not be reachable

--- Aborting in libhilti
/bin/sh: line 1: 39160 Abort trap: 6           spicy-driver -p Mini::Main /Users/bbannier/src/spicy/tests/.tmp/spicy.types.sink.reassembler.overlap/overlap.spicy > output < /dev/null

RESULT FILE .tmp/spicy.types.sink.reassembler.skip/.stderr:
[libhilti] Internal error: code is executing that should not be reachable

--- Aborting in libhilti
/bin/sh: line 1: 39290 Abort trap: 6           spicy-driver -p Mini::Main /Users/bbannier/src/spicy/tests/.tmp/spicy.types.sink.reassembler.skip/skip.spicy > output < /dev/null

RESULT FILE .tmp/spicy.types.unit.filter-chained/.stderr:
[libhilti] Internal error: code is executing that should not be reachable

--- Aborting in libhilti
/bin/sh: line 1: 44341 Done                    echo "H4sIAOVzEV0CA/NIzcnJ11EILshMrlQEACp6Q+YNAAAA"
     44342 Abort trap: 6           | spicy-driver -p Test::X /Users/bbannier/src/spicy/tests/.tmp/spicy.types.unit.filter-chained/filter-chained.spicy.hlto > output

RESULT FILE .tmp/spicy.types.unit.params-forward/.stderr:
[libhilti] Internal error: code is executing that should not be reachable

--- Aborting in libhilti
/bin/sh: line 1: 43385 Done                    printf xxx
     43386 Abort trap: 6           | spicy-driver -d -p Mini::Test /Users/bbannier/src/spicy/tests/.tmp/spicy.types.unit.params-forward/params-forward.spicy > output
```